test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+    this.d = {r};
+}
+
+// set up array of halo buffers over same distribution as 'u.targetLocales'
+var LOCALE_DOM = Block.createDomain(u.targetLocales().domain);
+var haloArrays: [LOCALE_DOM] [0..<4] haloArray;
+
+// buffer edge indices: North, East, South, West
+param N = 0, S = 1, E = 2, W = 3;
+
+// number of tasks per dimension based on Block distribution's decomposition
+const tidXMax = u.targetLocales().dim(0).high,
+      tidYMax = u.targetLocales().dim(1).high;
+
+// barrier for one task per locale
+var b = new barrier(numLocales);
+
+proc main() {
+  if RunCommDiag then startCommDiagnostics();
+
+  // spawn one task for each locale
+  t.start();
+  coforall (loc, (tidX, tidY)) in zip(u.targetLocales(), LOCALE_DOM) {
+    // run initialization and computation on the task for this locale
+    on loc {
+      // initialize halo arrays
+      haloArrays[tidX, tidY][N] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][S] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][E] = new haloArray(u.localSubdomain().dim(0).expand(1));
+      haloArrays[tidX, tidY][W] = new haloArray(u.localSubdomain().dim(0).expand(1));
+
+      // synchronize across tasks
+      b.barrier();
+
+      // run the portion of the FD computation owned by this task
+      work(tidX, tidY);
+    }
+  }
+
+  if RunCommDiag {
+    stopCommDiagnostics();
+    printCommDiagnosticsTable();
+  }
+
+  // print final results
+  const mean = (+ reduce u) / u.size,
+        stdDev = sqrt((+ reduce (u - mean)**2) / u.size);
+  t.stop();
+  writeln(abs(0.222751 - stdDev) < 1e-6);
+  if writeTime then writeln("time: ", t.elapsed());
+}
+
+proc work(tidX: int, tidY: int) {
+  // define domains to describe the indices owned by this task
+  const localIndices = u.localSubdomain(here),
+        localIndicesBuffered = localIndices.expand(1),
+        localIndicesInner = localIndices[indicesInner];
+
+	// declare two local arrays for computation on this locale
+  var uLocal1, uLocal2: [localIndicesBuffered] real = 1;
+
+  // populate local arrays with initial conditions from global array
+  uLocal1[localIndices] = u[localIndices];
+  uLocal2 = uLocal1;
+
+  // define constants for indexing into edges of local array
+  const WW = localIndicesBuffered.dim(1).low,
+        EE = localIndicesBuffered.dim(1).high,
+        SS = localIndicesBuffered.dim(0).high,
+        NN = localIndicesBuffered.dim(0).low;
+
+  // iterate for 'nt' time steps
+  for 1..nt {
+    // store results from last iteration in neighboring task's halo buffers
+    if tidY > 0       then haloArrays[tidX, tidY-1][E].v = uLocal2[.., WW+1];
+    if tidY < tidYMax then haloArrays[tidX, tidY+1][W].v = uLocal2[.., EE-1];
+    if tidX > 0       then haloArrays[tidX-1, tidY][S].v = uLocal2[NN+1, ..];
+    if tidX < tidXMax then haloArrays[tidX+1, tidY][N].v = uLocal2[SS-1, ..];
+
+    // swap local arrays
+    b.barrier();
+    uLocal1 <=> uLocal2;
```

Comment:
I'd suggest moving this swap somewhere else in this loop (maybe just after the following four population conditions?) so that someone just looking at the writes and reads of haloArrays won't be confused why we're reading from `uLocal2` but then writing to `uLocal1`.